### PR TITLE
CODENVY-739: pass proxy credentials to jgit

### PR DESCRIPTION
**Pull request by dmytro-ndp**
@vinokurig, @evoevodin: please, review this request.

Signed-off-by: Dmytro Nochevnov dnochevnov@codenvy.com

### CODE REVIEW

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 1]
```diff
@@ -235,6 +238,29 @@
     private final GitUserResolver   userResolver;
     private final Repository        repository;
 
+    // setup authenticators to use JGit behind the proxy
+    private static Authenticator httpsProxyAuth;
+    private static Authenticator httpProxyAuth;
+    static {
+        if (nonNull(System.getProperty("http.proxyUser")) && nonNull(System.getProperty("http.proxyPassword"))) {
```
Comment (voievodin) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: 1. I don't see any reason to use `nonNull(...)` against `!= null`.
2. I don't think that camel cased properties are good choice, i would  preferable change them to `http.proxy_user` and `http.proxy_password`
3. `Authenticator` looks not good as anonymous class to me, i think you  should extract it to the private static _NAMED_ inner class, and it seems that findbugs  will suggest you to do the same.

**Comment by voievodin** [comment]
reviewed, check the comment

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java (hunk 1 above)
Comment (dmytro-ndp) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: > (1). I don't see any reason to use nonNull(...) against != null.

Agree.

> (2). I don't think that camel cased properties are good choice, i would preferable change them to http.proxy_user and http.proxy_password

I don't think so. Camel case is good enough here, moreover - there is a convention of Java proxy settings naming which we can't change: http://rolandtapken.de/blog/2012-04/java-process-httpproxyuser-and-httpproxypassword

> (3). Authenticator looks not good as anonymous class to me, i think you should extract it to the private static NAMED inner class, and it seems that findbugs will suggest you to do the same.

Agree.

**Comment by dmytro-ndp** [comment]
Thank you! Please, see my answers.

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java (hunk 1 above)
Comment (vzhukovs) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Why did you use `java.util.Objects#nonNull` instead of `com.google.common.base.Strings#isNullOrEmpty`?
Empty property value may means that this property is not configured, isn't it?

Comment (dmytro-ndp) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Empty property shouldn't affect the code. But usage of _Strings#isNullOrEmpty_ is definitely great suggestion. Thank you!

**Comment by voievodin** [comment]
Well, then there is nothing to add. LGTM Good Job :+1:

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 2]
```diff
@@ -1590,13 +1616,22 @@ public void configure(Transport transport) {
                         sshTransport.setSshSessionFactory(sshSessionFactory);
                     }
                 });
+
+                if (nonNull(httpProxyAuth)) {
+                    Authenticator.setDefault(httpProxyAuth);
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: 1. Please, can you explain why we should use **http**ProxyAuth for **ssh** connections and **https**ProxyAuth for **http/https** ?
2. I guess there can be some problem with concurrency. I mean you can send two requests with ssh and http links. Then there can be situation when before invoking of `command.call()` will be invoked `Authenticator.setDefault(httpsProxyAuth);` for ssh link. Am I wrong?

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 3]
```diff
@@ -15,7 +15,6 @@
 import com.jcraft.jsch.JSch;
 import com.jcraft.jsch.JSchException;
 import com.jcraft.jsch.Session;
-
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java:18]: try to use formatter here. I guess it'll revert this change

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java (hunk 2 above)
Comment (dmytro-ndp) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: > (1) Please, can you explain why we should use httpProxyAuth for ssh connections and httpsProxyAuth for http/https ?

You found an error, and we even don't need setup proxy for ssh.

> (2) I guess there can be some problem with concurrency. I mean you can send two requests with ssh and http links. Then there can be situation when before invoking of command.call() will be invoked Authenticator.setDefault(httpsProxyAuth); for ssh link. Am I wrong?

You are right, but I would prefer don't work around synchronization of invoking authenticator so as it will lead to redundant delays.

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java (hunk 3 above)
Comment (dmytro-ndp) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java:18]: I will try. Thank you.

**Comment by dmytro-ndp** [comment]
@evoevodin, @sleshchenko: please, review my fixup on your notices.

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 4]
```diff
@@ -235,13 +235,16 @@
...
+    private final ProxyAuthenticator proxyAuthenticator;
+
     @Inject
     JGitConnection(Repository repository, CredentialsLoader credentialsLoader, SshKeyProvider sshKeyProvider,
                    GitUserResolver userResolver) {
         this.repository = repository;
         this.credentialsLoader = credentialsLoader;
         this.sshKeyProvider = sshKeyProvider;
         this.userResolver = userResolver;
+        this.proxyAuthenticator = new ProxyAuthenticator();
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: Seems that this instance is unused

Comment (dmytro-ndp) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: You are right.

File: wsagent/che-core-api-git/src/test/java/org/eclipse/che/git/impl/PullTest.java [hunk 5]
```diff
@@ -137,4 +137,29 @@ public void testWhenThereAreNoAnyRemotes(GitConnectionFactory connectionFactory)
         PullRequest request = newDto(PullRequest.class);
         connection.pull(request);
     }
+
+    @Test(dataProvider = "GitConnectionFactory", dataProviderClass = GitConnectionFactoryProvider.class,
+        expectedExceptions = GitException.class, expectedExceptionsMessageRegExp = "No remote repository specified.  " +
```
Comment (sleshchenko) [inline wsagent/che-core-api-git/src/test/java/org/eclipse/che/git/impl/PullTest.java:142]: Are you sure that this test check your use case?

**Comment by sleshchenko** [comment]
OK for me

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 6]
```diff
@@ -235,9 +237,8 @@
     private final GitUserResolver   userResolver;
     private final Repository        repository;
 
-    @Inject
-    JGitConnection(Repository repository, CredentialsLoader credentialsLoader, SshKeyProvider sshKeyProvider,
-                   GitUserResolver userResolver) {
+    @Inject JGitConnection(Repository repository, CredentialsLoader credentialsLoader, SshKeyProvider sshKeyProvider,
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: revert formatting

File: core/che-core-api-core/src/main/java/org/eclipse/che/commons/proxy/ProxyAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+ * https.proxyPassword
+ *
+ * Usage:
+ * ProxyAuthenticator.initAuthenticator(url)
+ * ... making http(s) request by url ...
+ * ProxyAuthenticator.resetAuthenticator()
+ *
+ * @author Dmytro Nochevnov
+ */
+public class ProxyAuthenticator extends Authenticator {
```
Comment (dmytro-ndp) [inline core/che-core-api-core/src/main/java/org/eclipse/che/commons/proxy/ProxyAuthenticator.java:36]: There is a test PullTest.

File: core/che-core-api-core/src/main/java/org/eclipse/che/commons/proxy/ProxyAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        currentProtocolHolder.remove();
+    }
+
+    private enum Protocol {
+        HTTP, HTTPS;
+
+        private PasswordAuthentication passwordAuthentication = createPasswordAuthentication();
+
+        private PasswordAuthentication createPasswordAuthentication() {
+            if (! (Strings.isNullOrEmpty(getProxyUserSystemProperty())
```
Comment (vinokurig) [inline core/che-core-api-core/src/main/java/org/eclipse/che/commons/proxy/ProxyAuthenticator.java]: Please use static import of `isNullOrEmpty`

File: wsagent/che-core-api-git/src/test/java/org/eclipse/che/git/impl/PullTest.java [hunk 9]
```diff
@@ -137,4 +137,29 @@ public void testWhenThereAreNoAnyRemotes(GitConnectionFactory connectionFactory)
...
+        System.setProperty("http.proxyUser", "user1");
+        System.setProperty("http.proxyPassword", "paswd1");
+        System.setProperty("https.proxyUser", "user2");
+        System.setProperty("https.proxyPassword", "paswd2");
+
+        GitConnection connection = connectToInitializedGitRepository(connectionFactory, repository);
+
+        //when
+        PullRequest request = newDto(PullRequest.class);
+        connection.pull(request);
```
Comment (vinokurig) [inline wsagent/che-core-api-git/src/test/java/org/eclipse/che/git/impl/PullTest.java:155]: Please make this in one line

File: core/che-core-api-core/src/main/java/org/eclipse/che/commons/proxy/ProxyAuthenticator.java (hunk 8 above)
Comment (dmytro-ndp) [inline core/che-core-api-core/src/main/java/org/eclipse/che/commons/proxy/ProxyAuthenticator.java]: Thank you for advice.

File: wsagent/che-core-api-git/src/test/java/org/eclipse/che/git/impl/PullTest.java (hunk 9 above)
Comment (dmytro-ndp) [inline wsagent/che-core-api-git/src/test/java/org/eclipse/che/git/impl/PullTest.java:155]: As for me, current variant is more readable.

**Comment by dmytro-ndp** [comment]
@evoevodin: please, review new fixup.

**Comment by voievodin** [comment]
LGTM

**Comment by dmytro-ndp** [comment]
Thank @evoevodin, @sleshchenko, @vinokurig for your helpful notices!
